src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import java.util
+
+import com.fulcrumgenomics.FgBioDef._
+import htsjdk.samtools.{Cigar, CigarElement, SAMRecord, SAMUtils, CigarOperator => Op}
+
+import scala.collection.JavaConversions.iterableAsScalaIterable
+import scala.collection.mutable.ListBuffer
+
+/**
+  *
+  */
+object SamRecordClipper {
+  /** An enumeration representing the various ways to clip bases within a read. */
+  object ClippingMode extends Enumeration { val Soft, Hard, SoftWithMask = Value }
+  type ClippingMode = ClippingMode.Value
+
+  private val NoCallBase = 'N'.toByte
+  private val NoCallQual = 2.toByte
+
+  // TODO: soft vs. hard vs. soft + masking
+  // TODO: start vs. end vs. 5' vs. 3' end of read
+  // TODO: bases to clip vs. ensure at least N clipped
+  // TODO: indel handling, soft-clip and hard-clip handling?
+  // TODO: handle unmapping the read if appropriate
+
+  /**
+    * Adds clipping of _at least_ numberOfBasesToClip to the start (left hand end)
+    * of an alignment.
+    *
+    * @param rec the record to be clipped
+    * @param numberOfBasesToClip the number of additional bases to clip beyond any clipping
+    *                            already applied to the read
+    */
+  def clipStartOfAlignment(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) : Unit = {
```

Comment:
For hard-clipping and soft-cliping-with-mask, it may be useful to return the bases and qualities that were removed or masked respectively.